File: backend/app/services/pricing.py
from decimal import ROUND_HALF_EVEN, Decimal

from app.models.base import Cents, Permille
from app.schemas.pricing import PriceBreakdown

PERMILLE = Decimal(1000)
WHOLE_CENT = Decimal(1)


def price_breakdown(net_cents: Cents, vat_rate_permille: Permille) -> PriceBreakdown:
    vat_cents = int(
        (Decimal(net_cents) * Decimal(vat_rate_permille) / PERMILLE).quantize(
            WHOLE_CENT, rounding=ROUND_HALF_EVEN
        )
    )
    return PriceBreakdown(net=net_cents, vat=vat_cents, gross=net_cents + vat_cents)

File: frontend/src/utils/pricing.ts
import type { PriceBreakdown } from "../orval/generated/fastAPI.schemas";

const PERMILLE = 1000;
const PERMILLE_PER_PERCENT = 10;

const divideHalfEven = (numerator: number, denominator: number) => {
  const sign = numerator < 0 ? -1 : 1;
  const dividend = Math.abs(numerator);
  const quotient = Math.floor(dividend / denominator);
  const doubledRemainder = (dividend % denominator) * 2;
  if (doubledRemainder > denominator) return sign * (quotient + 1);
  if (doubledRemainder < denominator) return sign * quotient;
  return sign * (quotient % 2 === 0 ? quotient : quotient + 1);
};

export const vatRatePermille = (vatRatePercent: number) =>
  Math.round(vatRatePercent * PERMILLE_PER_PERCENT);

export const vatCents = (netCents: number, vatRatePercent: number) =>
  divideHalfEven(netCents * vatRatePermille(vatRatePercent), PERMILLE);

export const priceBreakdown = (
  netCents: number,
  vatRatePercent: number,
): PriceBreakdown => {
  const vat = vatCents(netCents, vatRatePercent);
  return { net: netCents, vat, gross: netCents + vat };
};
